### PR TITLE
Allows plugin to register commands

### PR DESCRIPTION
**Pull request by Kingwl**
#43449

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@sheetalkamat this seems about right to me. Thoughts?

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
I would rather prefer if [enablePlugins ](https://github.com/microsoft/TypeScript/blob/master/src/server/project.ts#L1601) passes session (along with project which is already passed in) to the plugin rather than having these extra methods and we having to deal with the additions etc

**Comment by Kingwl** [comment]
I'm not pretty sure what's different about these kind of projects.

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -411,6 +411,7 @@ namespace ts.server {
         /** @deprecated use serverMode instead */
         syntaxOnly?: boolean;
         serverMode?: LanguageServiceMode;
+        session: Session<unknown> | undefined;
```
Comment (sheetalkamat) [inline src/server/editorServices.ts:414]: Please add type parameters to the options as well as Project service please with same default as session <TMessage = string>`

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: src/server/editorServices.ts [hunk 2]
```diff
@@ -783,10 +784,13 @@ namespace ts.server {
         /*@internal*/
         private packageJsonFilesMap: ESMap<Path, FileWatcher> | undefined;
 
+        /*@internal*/
+        readonly session: Session<unknown> | undefined;
```
Comment (sheetalkamat) [inline src/server/editorServices.ts:788]: I think you should wire TMessage thing here and PluginCreateInfo too,

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
Also add test for plugin where session is passed through and you call addProtocolHandler and test that functionality
Tests are in `src/testRunner/unittests/tsserver/plugins.ts`

File: src/server/editorServices.ts (hunk 2 above)
Comment (Kingwl) [reply src/server/editorServices.ts:788]: I thought the `unknown` is better:

If we add type parameter `TMessage ` into `editorService`, We should also add the type parameter into `Project` and etc. Too many changes.

And we are not really care about the type of message in `editorService`. We just forward session into pluin.

And the `handlers` or `addProtocolHandler` does not have `TMessage ` too. Thay are raw string.

Comment (sheetalkamat) [reply src/server/editorServices.ts:788]: If that’s the case we shouldn’t add it to project service options either

Comment (Kingwl) [reply src/server/editorServices.ts:788]: Agreed.

**Review by sheetalkamat — APPROVED** [review]
need to update tsserverlibrary.d.ts baseline but otherwise change looks good.  Thank you for your patience and sorry for going back and forth on this one as it was little unclear if TypeParameter was needed or not until you made those changes.

**Comment by Kingwl** [comment]
Never mind.

**Comment by RyanCavanaugh** [comment]
@Kingwl are you OK if I un-draft and merge this?

**Comment by Kingwl** [comment]
> @Kingwl are you OK if I un-draft and merge this?

Sure. Please feel free to merge.
Sorry. I havn't notice that this is a draft.

**Comment by sandersn** [comment]
Just noticed this in the PR backlog. I'm going to merge it since it looks ready to go.

**Pull request merged**